clang/test/SemaSYCL/intel-fpga-local-ast.cpp

```diff
@@ -0,0 +1,537 @@
+}
+
+template <int A>
+struct templ_st {
+  //CHECK: FieldDecl{{.*}}templ_force_p2d_field
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelForcePow2DepthAttr
+  //CHECK: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}0
+  //CHECK-NEXT: SubstNonTypeTemplateParmExpr
+  //CHECK-NEXT: NonTypeTemplateParmDecl
+  //CHECK-NEXT: IntegerLiteral{{.*}}0{{$}}
+  [[intel::force_pow2_depth(A)]] unsigned int templ_force_p2d_field[64];
+};
+
+int main() {
+  deviceQueue.submit([&](sycl::handler &h) {
+    h.single_task<class kernel_function>([]() {
+      check_ast();
+      //expected-note@+1{{in instantiation of function template specialization}}
+      check_template_parameters<2, 4, 8, 1>();
+      struct templ_st<0> ts {};
+    });
+  });
+
+  return 0;
+}
+
+// Test that checks global constant variable (which allows the redeclaration) since
+// SYCLIntelConstVar is one of the subjects listed for [[intel::max_replicates()]] attribute.
+
+// Check duplicate argument values with implicit memory attribute.
+//CHECK: VarDecl{{.*}}var_max_replicates
+//CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+//CHECK: SYCLIntelMaxReplicatesAttr
+//CHECK-NEXT: ConstantExpr
+//CHECK-NEXT: value:{{.*}}12
+//CHECK-NEXT: IntegerLiteral{{.*}}12{{$}}
+//CHECK: SYCLIntelMaxReplicatesAttr
+//CHECK-NEXT: ConstantExpr
+//CHECK-NEXT: value:{{.*}}12
+//CHECK-NEXT: IntegerLiteral{{.*}}12{{$}}
+[[intel::max_replicates(12)]] extern const int var_max_replicates;
+[[intel::max_replicates(12)]] const int var_max_replicates = 0; // OK
+
+// Merging of different arg values.
+//expected-warning@+2{{attribute 'max_replicates' is already applied with different arguments}}
+[[intel::max_replicates(12)]] extern const int var_max_replicates_1;
+[[intel::max_replicates(14)]] const int var_max_replicates_1 = 0;
+//expected-note@-2{{previous attribute is here}}
+
+// Test that checks global constant variable (which allows the redeclaration) since
+// IntelFPGAConstVar is one of the subjects listed for [[intel::force_pow2_depth()]] attribute.
+
+// Checking of duplicate argument values.
+//CHECK: VarDecl{{.*}}force_pow2_depth
+//CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+//CHECK: SYCLIntelForcePow2DepthAttr
+//CHECK-NEXT: ConstantExpr
+//CHECK-NEXT: value:{{.*}}1
+//CHECK-NEXT: IntegerLiteral{{.*}}1{{$}}
+//CHECK: SYCLIntelForcePow2DepthAttr
+//CHECK-NEXT: ConstantExpr
+//CHECK-NEXT: value:{{.*}}1
+//CHECK-NEXT: IntegerLiteral{{.*}}1{{$}}
+[[intel::force_pow2_depth(1)]] extern const int var_force_pow2_depth;
+[[intel::force_pow2_depth(1)]] const int var_force_pow2_depth = 0; // OK
+
+// Merging of different arg values.
+//expected-warning@+2{{attribute 'force_pow2_depth' is already applied with different arguments}}
+[[intel::force_pow2_depth(1)]] extern const int var_force_pow2_depth_1;
+[[intel::force_pow2_depth(0)]] const int var_force_pow2_depth_1 = 0;
+//expected-note@-2{{previous attribute is here}}
```

Comment:
For this _1 version of the variables and three that follow below, we are not checking ASTs, but rather just the diagnostic.
I think it makes sense to check the AST that it gets the first value (1).

Reply:
Done